### ISSUE TITLE
Not waiting for all data from socket before parsing message

### ISSUE THREAD

**Issue by ibash** [MatthieuLemoine/push-receiver#10]
Steps:
1. ??? I'm not sure what I did to get into this state (maybe have a bunch of gcm notifications queued?), but I'm consistently in it
2. Get multiple socket data events per gcm notification

Expected:
Able to parse notifications

Actual:
index out of range errors from protobuf

I think this is also the cause of #9 

When you get into this state, if you add a `console.log` in `socket.on('data'` like so:
`console.log('got data', buffer.length)`

Than for the first push notification you'll see the following:
```
got data 1
got data 88
got data 105
got data 514
got data 695
```

For subsequent push notifications you'll see the following:
```
got data 1026
got data 189
```

Note that the first push notification _always_ starts with a buffer of length 1.

When you're in this state `NotificationSchema.decode(buffer)` always fails with messages like:
```
Error: invalid wire type 7 at offset 9
    at BufferReader.Reader.skipType (/Users/islam/Projects/superhuman/desktop-electron/node_modules/protobufjs/src/reader.js:375:19)
    at BufferReader.Reader.skipType (/Users/islam/Projects/superhuman/desktop-electron/node_modules/protobufjs/src/reader.js:366:22)
    at Type.DataMessageStanza$decode [as decode] (eval at Codegen (/Users/islam/Projects/superhuman/desktop-electron/node_modules/@protobufjs/codegen/index.js:50:33), <anonymous>:67:5)
    at Type.decode_setup [as decode] (/Users/islam/Projects/superhuman/desktop-electron/node_modules/protobufjs/src/type.js:502:25)
    at onMessageReceived (/Users/islam/Projects/superhuman/desktop-electron/node_modules/push-receiver/src/client/socket/index.js:125:34)
    at Timeout.setTimeout [as _onTimeout] (/Users/islam/Projects/superhuman/desktop-electron/node_modules/push-receiver/src/client/socket/index.js:105:9)
    at ontimeout (timers.js:475:11)
    at tryOnTimeout (timers.js:310:5)
    at Timer.listOnTimeout (timers.js:270:5)
```

This is because `socket.on('data'` is not waiting for all the data to arrive before parsing it. I was able to demonstrate this by making a hack fix:

```
function listen(socket, NotificationSchema, keys, persistentIds) {
  let buf = null
  let timeout = null

  socket.on('data', (buffer) => {
    if (buf) {
      buf = Buffer.concat([buf, buffer])
    } else {
      buf = buffer
    }

    if (!timeout) {
      timeout = setTimeout(() => {
        onMessageReceived(buf, NotificationSchema, keys, persistentIds)
        buf = null
        timeout = null
      }, 1000)
    }
  });
}
```

This works but isn't the right fix since it's race-condition prone.

**Comment by ibash** [comment]
I think the correct way to do this is to copy what chromium does. It has an explicit sense of state and knows how many bytes to expect before processing the buffer. Check waitForData here: https://cs.chromium.org/chromium/src/google_apis/gcm/engine/connection_handler_impl.cc?rcl=dc7c41bc0ee5fee0ed269495dde6b8c40df43e40&l=178

**Comment by MatthieuLemoine** [comment]
@ibash Could you set up a PR so that we could try your solution ?

**Comment by ibash** [comment]
Will do - I wouldn't recommend it in production though.

**Comment by ibash** [comment]
See #11

**Comment by PedroKantar** [comment]
@ibash  Any news on this thread?

**Comment by ibash** [comment]
Hi @MatthieuLemoine and @PedroKantar! We're working on an implementation of superhuman using electron and I scheduled some time to take a pass at fixing this correctly. We probably won't get to it for a week or two though. (currently we're using the setTimeout hack as we flesh out different parts of the app).

**Comment by PedroKantar** [comment]
Thanks for your prompt answer..
Actually, I have also tested the setTimeout hack but based on the condition `0==persistentIds.length`. Because the issue is only on the very very first notification after a complete registration sequence (see #9). In this case `persistentIds` is set to `"persistentIds": []` in the config.json file.

**Comment by yurynix** [comment]
>Than for the first push notification you'll see the following:
got data 1
got data 88
got data 105
got data 514
got data 695

In my case,
The first thing I recv. when I connect is 1 byte MCS version -  41 (hex 29)
The next thing is Tag 3, `LoginResponse` ( 88 bytes in my case ) and then Tag 7 `IqStanza` ( 105 bytes in my case ).
Which are exactly the same numbers you have 1, 88, 105.
Those can be ignored, the next thing will be notification.

The code at #12 working for me, parsing an printing those if you wish to take a look.

**Comment by ibash** [comment]
@MatthieuLemoine / @yurynix FYI I started on this and I should have something usable in the next day or two. I'm copying the implementation from chromium (with a few slight simplifications) so hopefully it's mostly correct.

Also - @MatthieuLemoine would you be opposed to restructuring this library to use more classes? Ideally each connection can be isolated.

**Comment by MatthieuLemoine** [comment]
@ibash sure if it makes sense

**Comment by ibash** [comment]
Pull request is here: https://github.com/MatthieuLemoine/push-receiver/pull/13 let me know if you have questions / thoughts!

**Comment by MatthieuLemoine** [comment]
Fixed in v2.0.0
